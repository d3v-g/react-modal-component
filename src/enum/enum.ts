export enum ButtonType {
    PRIMARY,
    SECONDARY,
}

export enum ModalStatus {
    ERROR,
    SUCCESS,
}
