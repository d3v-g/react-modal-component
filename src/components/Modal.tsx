import { ButtonType, ModalStatus } from "../enum/enum"
import Button from "./Button"
import "../styles/components/modal.scss"

export interface IModalProps {
    isOpen: boolean
    title: string
    description: string
    status: ModalStatus
    onClose: () => void
    onContinue: () => void
}

export const Modal: React.FC<IModalProps> = ({
    isOpen,
    title,
    description,
    status,
    onClose,
    onContinue,
}) => {
    if (!isOpen) return null

    return (
        <div
            aria-labelledby="modal title"
            aria-describedby="modal description"
            aria-modal="true"
            className="modal__container"
            data-testid="modal"
        >
            <div className="modal__box" role="dialog" aria-label={title}>
                <div className="modal__content">
                    <h1
                        className={
                            "modal__title " +
                            (status === ModalStatus.SUCCESS
                                ? "modal__title--success"
                                : "modal__title--error")
                        }
                        id="modal title"
                    >
                        {title}
                    </h1>
                    <p id="modal description">{description}</p>
                </div>

                <hr className="modal__divider" />

                <div className="modal__children">
                    <Button
                        handleClick={onClose}
                        label="Cancel"
                        type={ButtonType.SECONDARY}
                    />
                    <Button
                        handleClick={onContinue}
                        label="Continue"
                        type={ButtonType.PRIMARY}
                    />
                </div>
            </div>
        </div>
    )
}

export default Modal
