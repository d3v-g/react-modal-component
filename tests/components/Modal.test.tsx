import React from "react"
import { render, screen } from "@testing-library/react"
import Modal, { IModalProps } from "../../src/components/Modal"
import userEvent, { UserEvent } from "@testing-library/user-event"
import { ModalStatus } from "../../src/enum/enum"
import { vi } from "vitest"

describe("Modal Component", () => {
    let modalProps: IModalProps

    describe("when modal is not open", () => {
        modalProps = {
            isOpen: false,
            title: "Test modal title",
            description: "Test description",
            status: ModalStatus.SUCCESS,
            onClose: vi.fn(),
            onContinue: vi.fn(),
        }

        it("does not render the modal", () => {
            render(<Modal {...modalProps} />)
            expect(screen.queryByTestId("modal")).not.toBeInTheDocument()
        })
    })

    describe("when modal is open", () => {
        beforeEach(() => {
            modalProps = {
                isOpen: true,
                title: "Test modal title",
                description: "Test description",
                status: ModalStatus.SUCCESS,
                onClose: vi.fn(),
                onContinue: vi.fn(),
            }
        })

        it("renders a modal", () => {
            render(<Modal {...modalProps} />)

            const modal = screen.getByTestId("modal")

            expect(modal).toBeInTheDocument()
            expect(modal).toHaveAttribute("aria-modal", "true")
        })

        it("renders a modal with the correct title", () => {
            const title = "render this title"
            modalProps = { ...modalProps, title }
            render(<Modal {...modalProps} />)

            const modalTitle = screen.getByRole("heading", { level: 1 })
            expect(screen.getByRole("dialog")).toHaveAccessibleName(title)
            expect(modalTitle).toHaveTextContent(title)

            expect(modalTitle).toHaveAttribute("id", "modal title")
            expect(screen.getByTestId("modal")).toHaveAttribute(
                "aria-labelledby",
                "modal title"
            )
        })

        it("renders a modal with the correct description", () => {
            const description = "render this description"
            modalProps = { ...modalProps, description }
            render(<Modal {...modalProps} />)

            const modalDescription = screen.getByText(description)

            expect(modalDescription).toBeInTheDocument()

            expect(modalDescription).toHaveAttribute("id", "modal description")
            expect(screen.getByTestId("modal")).toHaveAttribute(
                "aria-describedby",
                "modal description"
            )
        })

        describe("user interactions", () => {
            let user: UserEvent
            beforeEach(() => {
                user = userEvent.setup()
                modalProps = {
                    isOpen: true,
                    title: "Test modal title",
                    description: "Test description",
                    status: ModalStatus.SUCCESS,
                    onClose: vi.fn(),
                    onContinue: vi.fn(),
                }
            })

            test("onClose is called when cancel button is clicked", async () => {
                const onCloseSpy = vi.fn()
                modalProps = { ...modalProps, onClose: onCloseSpy }
                render(<Modal {...modalProps} />)

                const cancelButton = screen.getByRole("button", {
                    name: /cancel/i,
                })

                await user.click(cancelButton)
                expect(onCloseSpy).toHaveBeenCalledTimes(1)
            })

            test("onContinue is called when continue button is clicked", async () => {
                const onContinueSpy = vi.fn()
                modalProps = { ...modalProps, onContinue: onContinueSpy }
                render(<Modal {...modalProps} />)

                const cancelButton = screen.getByRole("button", {
                    name: /continue/i,
                })

                await user.click(cancelButton)
                expect(onContinueSpy).toHaveBeenCalledTimes(1)
            })
        })

        describe("appearance", () => {
            beforeEach(() => {
                modalProps = {
                    isOpen: true,
                    title: "Test modal title",
                    description: "Test description",
                    status: ModalStatus.SUCCESS,
                    onClose: vi.fn(),
                    onContinue: vi.fn(),
                }
            })

            test("modal with a success status", () => {
                const status = ModalStatus.SUCCESS
                modalProps = { ...modalProps, status }
                const { container } = render(<Modal {...modalProps} />)

                expect(container).toMatchSnapshot()
            })

            test("modal with an error status", () => {
                const status = ModalStatus.ERROR
                modalProps = { ...modalProps, status }
                const { container } = render(<Modal {...modalProps} />)

                expect(container).toMatchSnapshot()
            })
        })
    })
})
