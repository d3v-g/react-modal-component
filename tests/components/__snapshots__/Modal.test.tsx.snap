// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal Component > when modal is open > appearance > modal with a success status 1`] = `
<div>
  <div
    aria-describedby="modal description"
    aria-labelledby="modal title"
    aria-modal="true"
    class="modal__container"
    data-testid="modal"
  >
    <div
      aria-label="Test modal title"
      class="modal__box"
      role="dialog"
    >
      <div
        class="modal__content"
      >
        <h1
          class="modal__title modal__title--success"
          id="modal title"
        >
          Test modal title
        </h1>
        <p
          id="modal description"
        >
          Test description
        </p>
      </div>
      <hr
        class="modal__divider"
      />
      <div
        class="modal__children"
      >
        <button
          class="btn btn--secondary"
        >
          Cancel
        </button>
        <button
          class="btn btn--primary"
        >
          Continue
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal Component > when modal is open > appearance > modal with an error status 1`] = `
<div>
  <div
    aria-describedby="modal description"
    aria-labelledby="modal title"
    aria-modal="true"
    class="modal__container"
    data-testid="modal"
  >
    <div
      aria-label="Test modal title"
      class="modal__box"
      role="dialog"
    >
      <div
        class="modal__content"
      >
        <h1
          class="modal__title modal__title--error"
          id="modal title"
        >
          Test modal title
        </h1>
        <p
          id="modal description"
        >
          Test description
        </p>
      </div>
      <hr
        class="modal__divider"
      />
      <div
        class="modal__children"
      >
        <button
          class="btn btn--secondary"
        >
          Cancel
        </button>
        <button
          class="btn btn--primary"
        >
          Continue
        </button>
      </div>
    </div>
  </div>
</div>
`;
